experimental/dds/tree2/src/feature-libraries/editable-tree/editableTreeBinder.ts

```diff
@@ -322,6 +333,66 @@ abstract class AbstractPathVisitor implements PathVisitor {
 	protected readonly registeredListeners: Map<BindingContextType, Map<FieldKey, CallTree>> =
 		new Map();
 	public constructor(protected readonly options: BinderOptions) {}
+
+	// TODO: make these methods abstract and make AbstractPathVisitor implementations implement them
+	public afterCreate(content: DetachedRangeUpPath): void {}
+	public beforeDestroy(content: DetachedRangeUpPath): void {}
+	public beforeAttach(source: DetachedRangeUpPath, destination: PlaceUpPath): void {}
+	public afterAttach(source: DetachedPlaceUpPath, destination: RangeUpPath): void {
+		this.onInsert(
+			{
+				parent: destination.parent,
+				parentField: destination.field,
+				parentIndex: destination.start,
+			},
+			this.getContent(destination),
+		);
+	}
+	public beforeDetach(source: RangeUpPath, destination: DetachedPlaceUpPath): void {}
+	public afterDetach(source: PlaceUpPath, destination: DetachedRangeUpPath): void {
+		this.onDelete(
+			{
+				parent: source.parent,
+				parentField: source.field,
+				parentIndex: source.index,
+			},
+			destination.end - destination.start,
+		);
+	}
+	public beforeReplace(
+		newContent: DetachedRangeUpPath,
+		oldContent: RangeUpPath,
+		oldContentDestination: DetachedPlaceUpPath,
+		kind: ReplaceKind,
+	): void {}
+	public afterReplace(
+		newContentSource: DetachedPlaceUpPath,
+		newContent: RangeUpPath,
+		oldContent: DetachedRangeUpPath,
+		kind: ReplaceKind,
+	): void {
+		this.onDelete(
+			{
+				parent: newContent.parent,
+				parentField: newContent.field,
+				parentIndex: newContent.start,
+			},
+			oldContent.end - oldContent.start,
+		);
+		this.onInsert(
+			{
+				parent: newContent.parent,
+				parentField: newContent.field,
+				parentIndex: newContent.start,
+			},
+			this.getContent(newContent),
+		);
+	}
+	protected getContent(range: RangeUpPath): ProtoNodes {
+		// TODO: either lookup the content in the forest or stop providing the content in the events
+		return [];
```

Comment:
@dstanesc - This is a hack for the sake of moving the API forward. It will prevent change listeners from getting the real data (unless they look it up themselves). Is that fine to commit as is? Would you prefer that I update the `onInsert` API so it doesn't take the contents instead?

Reply:
@yann-achard-MS  - I believe your first proposal is best for converging to the final API faster. Please commit as is.